lib/msf/core/exploit/remote/ldap/active_directory.rb

```diff
@@ -0,0 +1,357 @@
+      # @return Returns nil when the object was not found.
+      # @rtype [Net::LDAP::Entry,nil]
+      def adds_get_object_by_dn(ldap, object_dn)
+        object = ldap_entry_cache.get_by_dn(object_dn)
+        return object if object
+
+        object = ldap.search(base: object_dn, controls: [adds_build_ldap_sd_control], scope: Net::LDAP::SearchScope_BaseObject)&.first
+        validate_query_result!(ldap.get_operation_result.table)
+
+        ldap_entry_cache << object if object
+        object
+      end
+
+      # Obtain a particular entity by its sAMAccountName.
+      #
+      # @param [Net::LDAP::Connection] ldap The LDAP connection to use for querying.
+      # @param [String] object_samaccountname The sAMAccountName of the object to retrieve.
+      # @return Returns nil when the object was not found.
+      # @rtype [Net::LDAP::Entry,nil]
+      def adds_get_object_by_samaccountname(ldap, object_samaccountname)
+        object = ldap_entry_cache.get_by_samaccountname(object_samaccountname)
+        return object if object
+
+        filter = "(sAMAccountName=#{ldap_escape_filter(object_samaccountname)})"
+        begin
+          object = ldap.search(base: ldap.base_dn, controls: [adds_build_ldap_sd_control], filter: filter)&.first
+        rescue Net::LDAP::Error => e
+          elog('ldap search error for sAMAccountName', error: e)
+          return nil
+        end
+        validate_query_result!(ldap.get_operation_result.table, filter)
+
+        ldap_entry_cache << object if object
+        object
+      end
+
+      # Obtain a particular entity by its SID.
+      #
+      # @param [Net::LDAP::Connection] ldap The LDAP connection to use for querying.
+      # @param [String] object_sid The SID of the object to retrieve.
+      # @return Returns nil when the object was not found.
+      # @rtype [Net::LDAP::Entry,nil]
+      def adds_get_object_by_sid(ldap, object_sid)
+        object_sid = Rex::Proto::MsDtyp::MsDtypSid.new(object_sid)
+        object = ldap_entry_cache.get_by_sid(object_sid)
+        return object if object
+
+        filter = "(objectSID=#{ldap_escape_filter(object_sid.to_s)})"
+        object = ldap.search(base: ldap.base_dn, controls: [adds_build_ldap_sd_control], filter: filter)&.first
+        validate_query_result!(ldap.get_operation_result.table, filter)
+
+        ldap_entry_cache << object if object
+        object
+      end
+
+      # Get the LDAP object that describes the current user.
+      #
+      # @param [Net::LDAP::Connection] ldap The LDAP connection to use for querying.
+      # @rtype [Net::LDAP::Entry]
+      def adds_get_current_user(ldap)
+        our_domain, _, our_username = ldap.ldapwhoami.to_s.delete_prefix('u:').partition('\\')
+        # todo: this is probably going to have issues if our user is from a domain that the target server is not the
+        # authority of
+        adds_get_object_by_samaccountname(ldap, our_username)
```

Comment:
Ah yes, that makes sense. 

What do you think about cacheing the result in an instance variable? The only reason I bring it up is I noticed `ldapwhoami` being called 144  times when running the `ldap_esc_vulnerable_cert_finder` module. 
```suggestion
        our_domain, _, @our_username = ldap.ldapwhoami.to_s.delete_prefix('u:').partition('\\') unless @our_username

        # todo: this is probably going to have issues if our user is from a domain that the target server is not the
        # authority of
        adds_get_object_by_samaccountname(ldap, @our_username)
```